tests/plugins/colorbutton/manual/colorhistory/4-colororder.md

```diff
@@ -0,0 +1,58 @@
+@bender-tags: 4.15.0, feature, 1795
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, colorbutton, colordialog, sourcearea, removeformat, undo
+
+**Note:** At the beginning open console and watch for errors (especially when picking colors).
+
+1. Click on color button.
+
+  **Expected:**
+
+  1. Below default palette there is horizontal rule and color history row.
+  1. There are 4 color boxes in color history row.
+  1. Color boxes are ordered by number of occurences and further by order in document:
+      * Strong Blue (occures 2 times in document)
+	  * Red (occures 1 time and is 1st color in doc)
+	  * Grey (occures 1 time and is 2nd color in doc)
+	  * Yellow (occures 1 time and is 3rd color in doc)
+
+  **Unexpected:**
+
+  * Any of the above conditions is not met.
+
+1. Click on `Automatic` color button.
+
+1. Click on color button.
+
+  **Expected:**
+
+  * There is no error in console.
+  * New color box didn't appear.
+
+  **Unexpected:**
+
+  * Empty color box appeared at the beginning of history or error occurred.
+
+1. Delete the entire editor contents.
+
+1. Click on color button.
+
+  **Expected:**
+
+  * Colors are still present and in the same order in history.
+
+  **Unexpected:**
+
+  * Colors disappeared or reordered.
+
+1. Click on background color button.
+
+  **Expected:**
+
+  * Color history row doesn't appear.
+
+  **Unexpected:**
+
+  * Color history is the same as for text color.
```

Comment:
Background color history row is visible as there are some background colors preloaded with the editor.

Reply:
At first I was going to write that you probably omitted the test step where you should clear editor, but in fact there is a way to produce a bug this way if where an empty span with background color set is left in the editor 🤔 I fixed this behaviour by filtering out empty spans.

Reply:
Indeed I omitted these test steps. Why colors are removed once the cleaning editor's content? If we consider this feature as a history register, shouldn't colors be preserved? AFAIK we discussed that it should work that way (I mean cleaning content shouldn't have an impact on color history).

Reply:
Indeed I omitted these test steps. Why colors are removed once the cleaning editor's content? If we consider this feature as a history register, shouldn't colors be preserved? AFAIK we discussed that it should work that way (I mean cleaning content shouldn't have an impact on color history).

Reply:
If a color appears in the history, it will not disappear. The content (and colors) are removed before first background color panel opening, and as color history is created on panel opening, there are no colors to be added to it - it's consistent with what we agreed on.

Reply:
> it's consistent with what we agreed on.

Not really. We agreed that color history should be filled with colors on editor initialization. So basically, you can reproduce something like:

* 1st user is defining some custom colors and saving it into a database.
* 2nd user is loading editor data and gets previously defined colors.

Do you see some use cases for the current implementation?

Reply:
Wow, really? I was pretty sure colors were about to load on panel initialization 🤔 As written [here](https://github.com/ckeditor/ckeditor4/pull/3784#issuecomment-576669938):
> Only at the first panel open (so onBlock) colors from content will be read and rendered in the history row.

I'm not quite sure what do you understand by saving the colors into a database and loading them by another user - should editor somehow keep those colors as a metadata? If user wants to suggest some colors to be used it can be done just by changing default palette, so I don't think we need to duplicate this feature 🤔 

The use case for the current implementation was:
* user opens editor;
* pastes content from external source (e.g. Word);
* opens color panel and has content colors in history to be easily picked up.

Reply:
> should editor somehow keep those colors as a metadata?

It already does - colors are preserved by content, right? So, opening color history on a styled content will load already defined colors by e.g. another user.

> The use case for the current implementation was:

I'm not sure if we should consider this use case, as it only works for the first-panel opening.

Reply:
Ok, as I misunderstood you - in a slack conversation we agreed that it's just about changing the moment when the colors from content is read - should be done on editor load, currently it is on panel load. I'll change it after current review round.